modules/dnn/src/layers/elementwise_layers.cpp

```diff
@@ -395,8 +395,29 @@ struct TanHFunctor
 #ifdef HAVE_OPENCL
     bool applyOCL(InputArrayOfArrays inps, OutputArrayOfArrays outs, OutputArrayOfArrays internals)
     {
-        // TODO: implement OCL version
-        return false;
+        size_t wgSize = ocl::Device::getDefault().maxWorkGroupSize();
+        std::vector<UMat> inputs;
+        std::vector<UMat> outputs;
+
+        inps.getUMatVector(inputs);
+        outs.getUMatVector(outputs);
+        String buildopt = oclGetTMacro(inputs[0]);
+
+        for (size_t i = 0; i < inputs.size(); i++)
+        {
+            UMat& src = inputs[i];
+            UMat& dst = outputs[i];
+
+            ocl::Kernel kernel("TanHForward", ocl::dnn::activations_oclsrc, buildopt);
+            kernel.set(0, (int)src.total());
+            kernel.set(1, ocl::KernelArg::PtrReadOnly(src));
+            kernel.set(2, ocl::KernelArg::PtrWriteOnly(dst));
+
+            size_t gSize = src.total();
+            CV_Assert(kernel.run(1, &gSize, &wgSize, false));
```

Comment:
I believe we should allow to OpenCL runtime to select appropriate group size:
```.diff
-&wgSize
+NULL
```

Any issues with this?

Reply:
No specific reason to use "wgSize", NULL is also OK in this patch. code is updated.